app/src/main/java/com/example/mycomifclient/connexion/ConnexionActivity.kt

```diff
@@ -1,22 +1,178 @@
+
         findViewById<Button>(R.id.a_connexion_button_connexion).setOnClickListener {
-            // TODO: Link this button to the connexion functionality
+            findViewById<Button>(R.id.a_connexion_button_connexion).isEnabled = false
+            id = this.findViewById<EditText>(R.id.a_connexion_edit_text_email).text.toString()
+            password =
+                this.findViewById<EditText>(R.id.a_connexion_edit_text_password).text.toString()
+            val authBody: JsonObject = createAuthBody(id, password)
+            authenticate(authBody)
         }
         findViewById<Button>(R.id.a_connexion_button_first_connexion).setOnClickListener {
+            findViewById<Button>(R.id.a_first_connexion_button_connexion).isEnabled = false
             val intent = Intent(this, FirstConnexionActivity::class.java)
             this.startActivity(intent)
         }
     }
+
+    override fun onResume() {
+        super.onResume()
+        findViewById<Button>(R.id.a_connexion_button_connexion).isEnabled = true
+        findViewById<Button>(R.id.a_connexion_button_first_connexion).isEnabled = true
+    }
+
+    private fun createAuthBody(username: String, password: String): JsonObject {
+        val serverBody = JsonObject()
+        serverBody.addProperty("client_id", 1)
+        serverBody.addProperty("client_secret", "secret")
+        serverBody.addProperty("grant_type", "password")
+        serverBody.addProperty("username", username)
+        serverBody.addProperty("password", password)
+        return serverBody
+    }
+
+    private fun authenticate(authBody: JsonObject) {
+
+        retrofitHTTPServices.authenticate(authBody).enqueue(object : Callback<JsonObject> {
+            override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
+                when (response.raw().code()) {
+
+                    200 -> handleAuthenticationResponse(response.body())
+
+                    401 -> Toast.makeText(
+                        this@ConnexionActivity,
+                        "Wrong credentials, please try again",
+                        Toast.LENGTH_LONG
+                    ).show()
+
+                    else -> println("Error")
+                }
+            }
+
+            override fun onFailure(call: Call<JsonObject>, t: Throwable) {
+                Toast.makeText(this@ConnexionActivity, "Error: $t", Toast.LENGTH_LONG).show()
+            }
+        })
+    }
+
+    private fun getUser() {
+        val user = userDAO.getFirst()
+        retrofitHTTPServices.getUser(218, "Bearer " + user.token)
+            .enqueue(object : Callback<JsonObject> {
+                override fun onResponse(
+                    call: Call<JsonObject>,
+                    response: Response<JsonObject>
+                ) {
+                    when (response.raw().code()) {
+
+                        200 -> handleGetUserResponse(response.body())
+
+                        401 -> reconnect()
```

Comment:
Why relaunch the activity ? I'd propose a simple toast message but I don't really understand what handles this 401 error.

Reply:
The main purpose of this is to handle 401 errors, which is due to a bad connexion/bad credentials. So when the error code is 401, you must reconnect. Yet I added a line that forces the MainActivity to stop when the error code is 401, which I forgot